server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java

```diff
@@ -87,19 +92,49 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
       );
     }
 
-    if (!outDir.mkdirs() && !outDir.isDirectory()) {
-      throw new IOException(String.format("Cannot create directory[%s]", outDir));
+    final File tmpOutDir = new File(baseStorageDir, intermediateDirFor(storageDir));
+    log.info("Creating intermediate directory[%s] for segment[%s]", tmpOutDir.toString(), segment.getIdentifier());
+    final long size = compressSegment(dataSegmentFile, tmpOutDir);
+
+    final DataSegment dataSegment = createDescriptorFile(
+      segment.withLoadSpec(makeLoadSpec(new File(outDir, "index.zip")))
+             .withSize(size)
+             .withBinaryVersion(SegmentUtils.getVersionFromDir(dataSegmentFile)),
+      tmpOutDir
+    );
+
+    // moving the temporary directory to the final destination, once success the potentially concurrent push operations
+    // will be failed and will read the descriptor.json created by current push operation directly
+    createIfNotExists(outDir.getParentFile());
+    try {
+      java.nio.file.Files.move(tmpOutDir.toPath(), outDir.toPath());
+    }
+    catch (FileAlreadyExistsException e) {
+      log.warn("Push destination directory[%s] exists, ignore this message if replication is configured.", outDir);
+      FileUtils.deleteDirectory(tmpOutDir);
+      return jsonMapper.readValue(new File(outDir, "descriptor.json"), DataSegment.class);
+    }
+    return dataSegment;
+  }
+
+  private void createIfNotExists(File directory) throws IOException
```

Comment:
suggest `createDirectoryIfNotExists` to make it more explicit in what its creating.